File: clients/ios/YunmaoApp/Sources/YunmaoApp/WebRTC/WhepClient.swift
// WhepClient：iOS WebRTC WHEP 拉流。
//
// 流程：
//   1. GET /v1/rooms/{id}/ice-servers      → IceServers
//   2. 创建 RTCPeerConnection（addTransceiver recvonly H.264 + Opus）
//   3. createOffer → setLocalDescription
//   4. POST /whep/{room_id} (application/sdp) → 收 answer
//   5. setRemoteDescription → 等 ICE/DTLS 完成
//   6. 渲染远端 video/audio track 到 RTCVideoRenderer。

import Foundation

public struct SdpValidationResult: Equatable {
    public let valid: Bool
    public let reason: String
    public init(valid: Bool, reason: String = "") {
        self.valid = valid
        self.reason = reason
    }
}

public enum SdpValidator {
    public static func validateAnswer(_ sdp: String) -> SdpValidationResult {
        if sdp.isEmpty { return SdpValidationResult(valid: false, reason: "SDP body is empty") }
        let lines = sdp.components(separatedBy: "\n")
        if !lines.contains(where: { $0.hasPrefix("v=") }) {
            return SdpValidationResult(valid: false, reason: "missing v= line")
        }
        if !lines.contains(where: { $0.hasPrefix("o=") }) {
            return SdpValidationResult(valid: false, reason: "missing o= line")
        }
        if !lines.contains(where: { $0.hasPrefix("s=") }) {
            return SdpValidationResult(valid: false, reason: "missing s= line")
        }
        let hasMedia = lines.contains(where: { $0.hasPrefix("m=video") || $0.hasPrefix("m=audio") })
        if !hasMedia {
            return SdpValidationResult(valid: false, reason: "no media section found")
        }
        return SdpValidationResult(valid: true)
    }

    public static func extractMediaTypes(_ sdp: String) -> [String] {
        sdp.components(separatedBy: "\n")
            .filter { $0.hasPrefix("m=") }
            .compactMap { line in
                let parts = line.dropFirst(2).split(separator: " ", maxSplits: 1)
                return parts.first.map(String.init)
            }
    }
}

#if canImport(WebRTC)
import WebRTC

public final class WhepClient: NSObject, RTCPeerConnectionDelegate {
    private let factory: RTCPeerConnectionFactory
    private var pc: RTCPeerConnection?
    public weak var videoRenderer: RTCVideoRenderer?

    public init(factory: RTCPeerConnectionFactory = .init()) {
        self.factory = factory
        super.init()
    }

    public func start(whepURL: URL, token: String, iceServers: [IceServersResponse.IceServer]) async throws {
        let config = RTCConfiguration()
        config.iceServers = iceServers.map { srv in
            RTCIceServer(urlStrings: srv.urls,
                         username: srv.username,
                         credential: srv.credential)
        }
        config.sdpSemantics = .unifiedPlan
        config.bundlePolicy = .maxBundle
        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        guard let pc = factory.peerConnection(with: config, constraints: constraints, delegate: self) else {
            throw NSError(domain: "whep", code: -1)
        }
        self.pc = pc

        let videoInit = RTCRtpTransceiverInit()
        videoInit.direction = .recvOnly
        _ = pc.addTransceiver(of: .video, init: videoInit)

        let audioInit = RTCRtpTransceiverInit()
        audioInit.direction = .recvOnly
        _ = pc.addTransceiver(of: .audio, init: audioInit)

        let offerConstraints = RTCMediaConstraints(mandatoryConstraints: [
            "OfferToReceiveAudio": "true",
            "OfferToReceiveVideo": "true",
        ], optionalConstraints: nil)
        let offer = try await withCheckedThrowingContinuation { (cont: CheckedContinuation<RTCSessionDescription, Error>) in
            pc.offer(for: offerConstraints) { sdp, error in
                if let error { cont.resume(throwing: error); return }
                guard let sdp else { cont.resume(throwing: NSError(domain: "whep", code: -2)); return }
                cont.resume(returning: sdp)
            }
        }
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            pc.setLocalDescription(offer) { error in
                if let error { cont.resume(throwing: error) } else { cont.resume() }
            }
        }

        var req = URLRequest(url: whepURL)
        req.httpMethod = "POST"
        req.setValue("application/sdp", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.httpBody = offer.sdp.data(using: .utf8)
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw NSError(domain: "whep", code: (resp as? HTTPURLResponse)?.statusCode ?? -3,
                          userInfo: ["body": String(data: data, encoding: .utf8) ?? ""])
        }
        guard let answerSDP = String(data: data, encoding: .utf8) else {
            throw NSError(domain: "whep", code: -4)
        }
        let validation = SdpValidator.validateAnswer(answerSDP)
        if !validation.valid {
            throw NSError(domain: "whep", code: -5,
                          userInfo: ["reason": validation.reason])
        }
        let answer = RTCSessionDescription(type: .answer, sdp: answerSDP)
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            pc.setRemoteDescription(answer) { error in
                if let error { cont.resume(throwing: error) } else { cont.resume() }
            }
        }
    }

    public func close() {
        pc?.close()
        pc = nil
    }

    // MARK: - RTCPeerConnectionDelegate

    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {}
    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {}
    public func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {}
    public func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {}
    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {}
    public func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {}
    public func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {}
    public func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}
    public func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {}
    public func peerConnection(_ peerConnection: RTCPeerConnection, didAdd rtpReceiver: RTCRtpReceiver, streams mediaStreams: [RTCMediaStream]) {
        if let track = rtpReceiver.track as? RTCVideoTrack, let renderer = videoRenderer {
            track.add(renderer)
        }
    }
}
#endif

File: clients/android/app/src/main/java/live/yunmao/app/webrtc/WhepClient.kt
package live.yunmao.app.webrtc

import android.content.Context
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.webrtc.IceCandidate
import org.webrtc.MediaConstraints
import org.webrtc.MediaStream
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.RtpReceiver
import org.webrtc.SdpObserver
import org.webrtc.SessionDescription
import org.webrtc.SurfaceViewRenderer

class WhepClient(
    context: Context,
) {
    private val factory: PeerConnectionFactory
    private var pc: PeerConnection? = null
    private val http = OkHttpClient.Builder().build()

    init {
        PeerConnectionFactory.initialize(
            PeerConnectionFactory.InitializationOptions.builder(context).createInitializationOptions()
        )
        factory = PeerConnectionFactory.builder().createPeerConnectionFactory()
    }

    fun start(
        whepUrl: String,
        token: String,
        iceServers: List<PeerConnection.IceServer>,
        renderer: SurfaceViewRenderer,
        onError: (Throwable) -> Unit,
    ) {
        val config = PeerConnection.RTCConfiguration(iceServers).apply {
            sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
        }
        val observer = object : PeerConnection.Observer {
            override fun onSignalingChange(state: PeerConnection.SignalingState?) {}
            override fun onIceConnectionChange(state: PeerConnection.IceConnectionState?) {}
            override fun onIceConnectionReceivingChange(p0: Boolean) {}
            override fun onIceGatheringChange(state: PeerConnection.IceGatheringState?) {}
            override fun onIceCandidate(c: IceCandidate?) {}
            override fun onIceCandidatesRemoved(p0: Array<out IceCandidate>?) {}
            override fun onAddStream(p0: MediaStream?) {}
            override fun onRemoveStream(p0: MediaStream?) {}
            override fun onDataChannel(p0: org.webrtc.DataChannel?) {}
            override fun onRenegotiationNeeded() {}
            override fun onAddTrack(receiver: RtpReceiver?, mediaStreams: Array<out MediaStream>?) {
                val track = receiver?.track()
                if (track is org.webrtc.VideoTrack) {
                    track.addSink(renderer)
                }
            }
        }
        val pc = factory.createPeerConnection(config, observer) ?: run {
            onError(IllegalStateException("createPeerConnection returned null"))
            return
        }
        this.pc = pc
        pc.addTransceiver(
            org.webrtc.MediaStreamTrack.MediaType.MEDIA_TYPE_VIDEO,
            org.webrtc.RtpTransceiver.RtpTransceiverInit(org.webrtc.RtpTransceiverDirection.RECV_ONLY)
        )
        pc.addTransceiver(
            org.webrtc.MediaStreamTrack.MediaType.MEDIA_TYPE_AUDIO,
            org.webrtc.RtpTransceiver.RtpTransceiverInit(org.webrtc.RtpTransceiverDirection.RECV_ONLY)
        )
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
        }
        pc.createOffer(object : SimpleSdpObserver() {
            override fun onCreateSuccess(offer: SessionDescription) {
                pc.setLocalDescription(object : SimpleSdpObserver() {
                    override fun onSetSuccess() {
                        Thread {
                            try {
                                val req = Request.Builder().url(whepUrl)
                                    .addHeader("Content-Type", "application/sdp")
                                    .addHeader("Authorization", "Bearer $token")
                                    .post(offer.description.toRequestBody("application/sdp".toMediaTypeOrNull()))
                                    .build()
                                val resp = http.newCall(req).execute()
                                if (!resp.isSuccessful) {
                                    onError(IllegalStateException("WHEP ${resp.code}"))
                                    return@Thread
                                }
                                val answerSdp = resp.body!!.string()
                                val validation = SdpValidator.validateAnswer(answerSdp)
                                if (!validation.valid) {
                                    onError(IllegalStateException("Invalid WHEP answer SDP: ${validation.reason}"))
                                    return@Thread
                                }
                                val answer = SessionDescription(SessionDescription.Type.ANSWER, answerSdp)
                                pc.setRemoteDescription(SimpleSdpObserver(), answer)
                            } catch (e: Throwable) {
                                onError(e)
                            }
                        }.start()
                    }
                }, offer)
            }
        }, constraints)
    }

    fun close() {
        pc?.close()
        pc = null
    }
}

data class SdpValidationResult(val valid: Boolean, val reason: String = "")

object SdpValidator {
    fun validateAnswer(sdp: String): SdpValidationResult {
        if (sdp.isBlank()) return SdpValidationResult(false, "SDP body is empty")
        val lines = sdp.lines()
        val hasVersion = lines.any { it.startsWith("v=") }
        if (!hasVersion) return SdpValidationResult(false, "missing v= line")
        val hasOrigin = lines.any { it.startsWith("o=") }
        if (!hasOrigin) return SdpValidationResult(false, "missing o= line")
        val hasSessionName = lines.any { it.startsWith("s=") }
        if (!hasSessionName) return SdpValidationResult(false, "missing s= line")
        val hasMediaVideo = lines.any { it.startsWith("m=video") }
        val hasMediaAudio = lines.any { it.startsWith("m=audio") }
        if (!hasMediaVideo && !hasMediaAudio) return SdpValidationResult(false, "no media section found")
        return SdpValidationResult(true)
    }

    fun extractMediaTypes(sdp: String): List<String> {
        return sdp.lines()
            .filter { it.startsWith("m=") }
            .map { it.substring(2).split(" ").firstOrNull() ?: "" }
    }
}

open class SimpleSdpObserver : SdpObserver {
    override fun onCreateSuccess(p0: SessionDescription?) {}
    override fun onSetSuccess() {}
    override fun onCreateFailure(p0: String?) {}
    override fun onSetFailure(p0: String?) {}
}

    fun start(
        whepUrl: String,
        token: String,
        iceServers: List<PeerConnection.IceServer>,
        renderer: SurfaceViewRenderer,
        onError: (Throwable) -> Unit,
    ) {
        val config = PeerConnection.RTCConfiguration(iceServers).apply {
            sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
        }
        val observer = object : PeerConnection.Observer {
            override fun onSignalingChange(state: PeerConnection.SignalingState?) {}
            override fun onIceConnectionChange(state: PeerConnection.IceConnectionState?) {}
            override fun onIceConnectionReceivingChange(p0: Boolean) {}
            override fun onIceGatheringChange(state: PeerConnection.IceGatheringState?) {}
            override fun onIceCandidate(c: IceCandidate?) {}
            override fun onIceCandidatesRemoved(p0: Array<out IceCandidate>?) {}
            override fun onAddStream(p0: MediaStream?) {}
            override fun onRemoveStream(p0: MediaStream?) {}
            override fun onDataChannel(p0: org.webrtc.DataChannel?) {}
            override fun onRenegotiationNeeded() {}
            override fun onAddTrack(receiver: RtpReceiver?, mediaStreams: Array<out MediaStream>?) {
                val track = receiver?.track()
                if (track is org.webrtc.VideoTrack) {
                    track.addSink(renderer)
                }
            }
        }
        val pc = factory.createPeerConnection(config, observer) ?: run {
            onError(IllegalStateException("createPeerConnection returned null"))
            return
        }
        this.pc = pc
        pc.addTransceiver(
            org.webrtc.MediaStreamTrack.MediaType.MEDIA_TYPE_VIDEO,
            org.webrtc.RtpTransceiver.RtpTransceiverInit(org.webrtc.RtpTransceiver.RtpTransceiverDirection.RECV_ONLY)
        )
        pc.addTransceiver(
            org.webrtc.MediaStreamTrack.MediaType.MEDIA_TYPE_AUDIO,
            org.webrtc.RtpTransceiver.RtpTransceiverInit(org.webrtc.RtpTransceiver.RtpTransceiverDirection.RECV_ONLY)
        )
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
        }
        pc.createOffer(object : SimpleSdpObserver() {
            override fun onCreateSuccess(offer: SessionDescription) {
                pc.setLocalDescription(object : SimpleSdpObserver() {
                    override fun onSetSuccess() {
                        Thread {
                            try {
                                val req = Request.Builder().url(whepUrl)
                                    .addHeader("Content-Type", "application/sdp")
                                    .addHeader("Authorization", "Bearer $token")
                                    .post(offer.description.toRequestBody("application/sdp".toMediaTypeOrNull()))
                                    .build()
                                val resp = http.newCall(req).execute()
                                if (!resp.isSuccessful) {
                                    onError(IllegalStateException("WHEP ${'$'}{resp.code}"))
                                    return@Thread
                                }
                                val answer = SessionDescription(SessionDescription.Type.ANSWER, resp.body!!.string())
                                pc.setRemoteDescription(SimpleSdpObserver(), answer)
                            } catch (e: Throwable) {
                                onError(e)
                            }
                        }.start()
                    }
                }, offer)
            }
        }, constraints)
    }

    fun close() {
        pc?.close()
        pc = null
    }
}

open class SimpleSdpObserver : SdpObserver {
    override fun onCreateSuccess(p0: SessionDescription?) {}
    override fun onSetSuccess() {}
    override fun onCreateFailure(p0: String?) {}
    override fun onSetFailure(p0: String?) {}
}
